### PR TITLE
chore: update RustPython depend

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Update RustPython Version(with GC on) and fix compile errors

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: py_time_poc.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+time:
+datetime(str, format= "%Y-%m-%d %H:%M:%S")->PyDateTime: 从字符串中解析日期时间
+默认ymd hms, 也可以自定义格式
```
Comment (killme2008) [inline py_time_poc.md:3]: what is this file for?

File: src/script/Cargo.toml [hunk 2]
```diff
@@ -46,17 +46,17 @@ futures-util.workspace = true
...
 paste = { workspace = true, optional = true }
 query = { path = "../query" }
-# TODO(discord9): This is a forked and tweaked version of RustPython, please update it to newest original RustPython After Update toolchain to 1.65
-rustpython-ast = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345" }
-rustpython-codegen = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345" }
-rustpython-compiler = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345" }
-rustpython-compiler-core = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345" }
-rustpython-parser = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345" }
-rustpython-pylib = { git = "https://github.com/discord9/RustPython", optional = true, rev = "2e126345", features = [
+# TODO(discord9): This is a forked and tweaked version of RustPython, please update it to newest original RustPython After RustPython support GC
```
Comment (killme2008) [inline src/script/Cargo.toml:49]: What's the status of the GC PR in rustpython? Are u still working on it?

File: src/script/src/python/ffi_types/vector/tests.rs [hunk 3]
```diff
@@ -155,8 +155,10 @@ fn eval_pyo3(testcase: TestCase, locals: HashMap<String, PyVector>) {
 }
 
 fn eval_rspy(testcase: TestCase, locals: HashMap<String, PyVector>) {
-    vm::Interpreter::without_stdlib(Default::default()).enter(|vm| {
+    vm::Interpreter::with_init(Default::default(), |vm| {
```
Comment (killme2008) [inline src/script/src/python/ffi_types/vector/tests.rs:158]: We don't want the stdlib, so does `with_init` create the interpreter with stdlib or not?

File: src/script/src/python/ffi_types/vector/tests.rs [hunk 4]
```diff
@@ -169,7 +171,14 @@ fn eval_rspy(testcase: TestCase, locals: HashMap<String, PyVector>) {
             .compile(&testcase.eval, Mode::Eval, "<embedded>".to_owned())
```
Comment (killme2008) [inline src/script/src/python/ffi_types/vector/tests.rs:171]:
```suggestion
            .compile(&testcase.eval, Mode::Eval, "<embedded>".to_string())
```

File: src/script/src/python/rspython/dataframe_impl.rs [hunk 5]
```diff
@@ -313,7 +316,20 @@ pub(crate) mod data_frame {
         }
     }
 
-    #[rspyclass(with(Comparable))]
+    impl AsNumber for PyExpr {
+        fn as_number() -> &'static PyNumberMethods {
+            static AS_NUMBER: PyNumberMethods = PyNumberMethods {
+                and: Some(|a, b, vm| PyExpr::and(a.to_owned(), b.to_owned(), vm).to_pyresult(vm)),
```
Comment (killme2008) [inline src/script/src/python/rspython/dataframe_impl.rs:322]: Please don't use `to_owned` any more.

File: src/script/src/python/rspython/utils.rs [hunk 6]
```diff
@@ -33,7 +34,21 @@ use crate::python::rspython::builtins::try_into_columnar_value;
 /// use `rustpython`'s `is_instance` method to check if a PyObject is a instance of class.
 /// if `PyResult` is Err, then this function return `false`
 pub fn is_instance<T: PyPayload>(obj: &PyObjectRef, vm: &VirtualMachine) -> bool {
-    obj.is_instance(T::class(vm).into(), vm).unwrap_or(false)
+    obj.is_instance(T::class(&vm.ctx).into(), vm)
```
Comment (killme2008) [inline src/script/src/python/rspython/utils.rs:37]: Duplicated with `src/script/src/python/utils.rs`, can we remove one?

File: src/script/src/python/rspython/utils.rs [hunk 7]
```diff
@@ -33,7 +34,21 @@ use crate::python::rspython::builtins::try_into_columnar_value;
...
+        .unwrap_or(false)
+}
+
+pub fn obj_cast_to<T: PyObjectPayload>(
+    obj: PyObjectRef,
+    vm: &VirtualMachine,
+) -> PyResult<PyRef<T>> {
+    obj.downcast::<T>().map_err(|e| {
+        vm.new_type_error(format!(
+            "Can't cast right operand into {}, actual type: {}",
```
Comment (killme2008) [inline src/script/src/python/rspython/utils.rs:47]: Maybe it's not the right operand? The error message needs to be corrected.

File: src/script/src/python/rspython/vector_impl.rs [hunk 8]
```diff
@@ -194,19 +203,33 @@ impl PyVector {
         }
     }
 
+    fn obj_to_vector(obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyRef<PyVector>> {
```
Comment (killme2008) [inline src/script/src/python/rspython/vector_impl.rs:206]: Duplicated with `obj_cast_to`.

File: src/script/src/python/rspython/vector_impl.rs [hunk 9]
```diff
@@ -262,6 +285,38 @@ impl PyVector {
...
+impl Representable for PyVector {
+    #[inline]
+    fn repr_str(zelf: &Py<Self>, _vm: &VirtualMachine) -> PyResult<String> {
+        Ok(format!("{:#?}", *zelf))
+    }
+}
+
+impl AsNumber for PyVector {
+    fn as_number() -> &'static PyNumberMethods {
+        static AS_NUMBER: PyNumberMethods = PyNumberMethods {
```
Comment (killme2008) [inline src/script/src/python/rspython/vector_impl.rs:297]: Again, don't use `to_owned` at all if you have other choices.

File: src/script/Cargo.toml (hunk 2 above)
Comment (discord9) [reply src/script/Cargo.toml:49]: Almost done, the maintainer is currently breaking GC PR into two(First a `Traverse` feature then core of GC algorithm and currently reviewing `Traverse` PR)https://github.com/RustPython/RustPython/pull/4872

> What's the status of the GC PR in rustpython? Are u still working on it?

File: src/script/src/python/ffi_types/vector/tests.rs (hunk 3 above)
Comment (discord9) [reply src/script/src/python/ffi_types/vector/tests.rs:158]: > We don't want the stdlib, so does `with_init` create the interpreter with stdlib or not?

This wouldn't init stdlib unless explictly write `vm.add_native_modules` etc.

File: src/script/src/python/rspython/vector_impl.rs (hunk 9 above)
Comment (discord9) [reply src/script/src/python/rspython/vector_impl.rs:297]: > Again, don't use `to_owned` at all if you have other choices.

It seems the devs have attach some semantics to `to_owned`... it create a new Ref and inc ref cnt by one, and there seems not having a better way to do this convert other than a `to_owned`, maybe it's a wrong api design here?
```
impl ToOwned for PyObject {
    type Owned = PyObjectRef;

    #[inline(always)]
    fn to_owned(&self) -> Self::Owned {
        self.0.ref_count.inc();
        PyObjectRef {
            ptr: NonNull::from(self),
        }
    }
}
```

Comment (killme2008) [reply src/script/src/python/rspython/vector_impl.rs:299]: Got it.  So there is no other choice, use `to_owned` is fine here.

File: src/script/Cargo.toml (hunk 2 above)
Comment (killme2008) [reply src/script/Cargo.toml:49]: Cool, great news.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
Have skimmed this and it generally looks good to me 👍

**Pull request merged**